open_mastr/xml_download/utils_download_bulk.py

```diff
@@ -5,30 +5,73 @@
 
 import numpy as np
 import requests
-from bs4 import BeautifulSoup
 from tqdm import tqdm
 
 # setup logger
 from open_mastr.utils.config import setup_logger
 
 log = setup_logger()
 
+def gen_version(when: time.struct_time = time.localtime()) -> str:
+    """
+    Generates the current version.
+
+    The version number is determined according to a fixed release cycle,
+    which is by convention in sync with the changes to other german regulatory
+    frameworks of the energysuch as GeLI Gas and GPKE.
+
+    The release schedule is twice per year on 1st of April and October.
+    The version number is determined by the year of release and the running
+    number of the release, i.e. the release on April 1st is release 1,
+    while the release in October is release 2.
+
+    Further, the release happens during the day, so on the day of the
+    changeover, the exported data will still be in the old version/format.
+
+    see <https://www.marktstammdatenregister.de/MaStRHilfe/files/webdienst/Release-Termine.pdf>
+
+    Examples:
+    2024-01-01 = version 23.2
+    2024-04-01 = version 23.2
+    2024-04-02 = version 24.1
+    2024-09-30 = version 24.1
+    2024-10-01 = version 24.1
+    2024-10-02 = version 24.2
+    2024-31-12 = version 24.2
+    """
 
-def get_url_from_Mastr_website() -> str:
-    """Get the url of the latest MaStR file from markstammdatenregister.de.
+    year = when.tm_year
+    release = 1
 
-    The file and the corresponding url are updated once per day.
-    The url has a randomly generated string appended, so it has to be
-    grabbed from the marktstammdatenregister.de homepage.
-    For further details visit https://www.marktstammdatenregister.de/MaStR/Datendownload
+    if when.tm_mon < 4 or (when.tm_mon == 4 and when.tm_mday == 1):
+        year = year - 1
+        release = 2
+    elif when.tm_mon > 10 or (when.tm_mon == 10 and when.tm_mday > 1):
+        release = 2
+
+    # only the last two digits of the year are used
+    year = str(year)[-2:]
+
+    return f'{year}.{release}'
+
+def gen_url(when: time.struct_time = time.localtime()) -> str:
```

Comment:
First of all thanks for this very useful PR @Johann150 
One remark here: I think it would be great to have a fallback url for the case that the current data is not online yet. This could happen, as you described, before 04:00.
Maybe we do the following: When the url is used to download the "Gesamtdatenexport" this is wrapped in a `try` - `except` block. If it fails, the url is changed to the url from one day before and the download is started again?

What do you think? And do you have time to make this change? Otherwise I can also do it.

Reply:
I have made the change. It was previously not checked at all what the download request status code was, so instead of writing a giant `try` block I thought it would be a better idea to "just" check the status code instead.

There is one potential situation that I'm not quite sure about, when thinking about this a bit more. Since the download can take a few minutes, if by coincidence you were to start the download right before the new file is published, I don't know if the rest of the old file will be correctly downloaded. But maybe that is a very hypothetical and contrived situation that does not really need to be considered.